wasp/apps/alarm.py

```diff
@@ -0,0 +1,155 @@
+# SPDX-License-Identifier: LGPL-3.0-or-later
+# Copyright (C) 2020 Daniel Thompson
+# Copyright (C) 2020 Joris Warmbier
+"""Alarm Application
+~~~~~~~~~~~~~~~~~~~~
+
+An application to set a vibration alarm. All settings can be accessed from the Watch UI
+"""
+
+import wasp
+import icons
+import time
+
+icon = (
+    96, 64,
+    b'\xff\x00\xff\x00\xbf\x05\x1f\x055\x08\x0b\x07\x0b\t1\n'
+    b'\x06\x0f\x07\t0\n\x06\x11\x06\n.\n\x04\x17\x04\n'
+    b'-\t\x04\x19\x04\t,\t\x03\x1e\x02\t+\x08\x03 '
+    b'\x02\x08+\x07\x03\r\x07\x0e\x02\x07+\x06\x03\r\t\r'
+    b'\x03\x06+\x05\x03\x0b\x0f\x0b\x03\x05+\x04\x03\n\x13\n'
+    b'\x03\x04,\x02\x04\t\x15\t\x04\x022\t\x17\t6\t'
+    b'\x19\t4\n\x1a\t3\t\r\x01\x0e\x083\x07\x0e\x03'
+    b'\x0e\x072\x08\x0e\x03\x0e\x081\x07\x0f\x03\x0f\x071\x07'
+    b'\x0f\x03\x0f\x071\x07\x0f\x03\x0f\x071\x06\x10\x03\x10\x06'
+    b'0\x07\x10\x03\x10\x07/\x07\x10\x03\x10\x07/\x07\x10\x03'
+    b'\x10\x07/\x07\x0f\x04\x10\x07/\x07\x0e\x04\x11\x07/\x07'
+    b'\r\x04\x12\x070\x06\x0c\x04\x13\x061\x07\n\x04\x13\x07'
+    b'1\x07\n\x03\x14\x071\x07\n\x02\x15\x071\x08\x1f\x08'
+    b'2\x07\x1f\x073\x08\x1d\x083\t\x1b\t4\t\x19\t'
+    b'6\t\x17\t8\t\x15\t9\n\x13\n:\x0b\x0f\x0b'
+    b'<\x0e\x08\r>!@\x1fB\x1dC\x1e@!?\x05'
+    b'\x03\x11\x03\x05?\x04\t\x07\t\x04@\x02\x0b\x05\x0b\x02'
+    b'\xff\x00\xff\x00a'
+)
+
+class AlarmApp():
+    """Allows the user to set a vibration alarm.
+    """
+    NAME = 'Alarm'
+    ICON = icon
+
+    def __init__(self):
+        """Initialize the application."""
+
+        self.active = False
+        self.ringing = False
+        self.hours = 0
+        self.minutes = 0
+
+        self._set_current_alarm()
+
+    def foreground(self):
+        """Activate the application."""
+        self._draw()
+        wasp.system.request_event(wasp.EventMask.TOUCH)
+        if self.ringing:
+            wasp.system.request_tick(1000)
+
+    def background(self):
+        """De-activate the application."""
+        wasp.system.cancel_alarm(time.mktime(self.current_alarm), self._alert)
```

Comment:
I think it might be better to cancel the alarm from the foreground() method. Seems needlessly difficult to test the app if the callback can trigger when it is already running.